cmd/fs-v1-multipart.go

```diff
@@ -726,248 +475,162 @@ func (fs fsObjects) ListObjectParts(bucket, object, uploadID string, partNumberM
+			partSize = fi.Size()
+		}
+		if i == len(parts)-1 {
+			break
 		}
 
 		// All parts except the last part has to be atleast 5MB.
-		if (i < len(parts)-1) && !isMinAllowedPartSize(fsMeta.Parts[partIdx].Size) {
-			fs.rwPool.Close(fsMetaPathMultipart)
+		if !isMinAllowedPartSize(fi.Size()) {
 			return oi, errors.Trace(PartTooSmall{
 				PartNumber: part.PartNumber,
-				PartSize:   fsMeta.Parts[partIdx].Size,
+				PartSize:   fi.Size(),
 				PartETag:   part.ETag,
 			})
 		}
-		if partSize == -1 {
-			partSize = fsMeta.Parts[partIdx].Size
-		}
+
 		// TODO: Make necessary changes in future as explained in the below comment.
 		// All parts except the last part has to be of same size. We are introducing this
 		// check to see if any clients break. If clients do not break then we can optimize
 		// multipart PutObjectPart by writing the part at the right offset using pwrite()
 		// so that we don't need to do background append at all. i.e by the time we get
 		// CompleteMultipartUpload we already have the full file available which can be
 		// renamed to the main name-space.
-		if (i < len(parts)-1) && partSize != fsMeta.Parts[partIdx].Size {
-			fs.rwPool.Close(fsMetaPathMultipart)
+		if partSize != fi.Size() {
 			return oi, errors.Trace(PartsSizeUnequal{})
 		}
 	}
 
-	// Wait for any competing PutObject() operation on bucket/object, since same namespace
-	// would be acquired for `fs.json`.
-	fsMetaPath := pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket, object, fsMetaJSONFile)
-	metaFile, err := fs.rwPool.Create(fsMetaPath)
-	if err != nil {
-		fs.rwPool.Close(fsMetaPathMultipart)
-		return oi, toObjectErr(errors.Trace(err), bucket, object)
-	}
-	defer metaFile.Close()
-
-	fsNSObjPath := pathJoin(fs.fsPath, bucket, object)
-
-	// This lock is held during rename of the appended tmp file to the actual
-	// location so that any competing GetObject/PutObject/DeleteObject do not race.
 	appendFallback := true // In case background-append did not append the required parts.
-
-	if isPartsSame(fsMeta.Parts, parts) {
-		err = fs.complete(bucket, object, uploadID, fsMeta)
-		if err == nil {
-			appendFallback = false
-			fsTmpObjPath := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, uploadID)
-			if err = fsRenameFile(fsTmpObjPath, fsNSObjPath); err != nil {
-				fs.rwPool.Close(fsMetaPathMultipart)
-				return oi, toObjectErr(err, minioMetaTmpBucket, uploadID)
+	appendFilePath := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, fmt.Sprintf("%s.%s", uploadID, mustGetUUID()))
+
+	// Most of the times appendFile would already be fully appended by now. We call fs.backgroundAppend()
+	// to take care of the following corner case:
+	// 1. The last PutObjectPart triggers go-routine fs.backgroundAppend, this go-routine has not started yet.
+	// 2. Now CompleteMultipartUpload gets called which sees that lastPart is not appended and starts appending
+	//    from the beginning
+	fs.backgroundAppend(bucket, object, uploadID)
```

Comment:
Should this still be called backgroundAppend() ?

Reply:
Yes, because it still appends in the background (though it is not in a long lasting go-routine listening on a channel)